### PR TITLE
Automate velero secret deployment

### PR DESCRIPTION
**Pull request by ShyamsundarR**
By adding a new policy and placement to deliver the secret from
the ramen namespace on the managed cluster to the velero namespace
on the managed cluster.

Using a single policy fails as that expects to be able to resolve
the configuration policy before applying the same. As the velero
secret is delivered using the delivered S3 secret in the ramen
namespace, such a single policy scheme fails.

Related changes to protect the secret using additional finalizers
and addressing upgrade cases are also covered in the commit.

Further to delivering the secret, update s3StoreAccessorsGet to process
the new secret as the veleroNamespaceSecretKeyRef when the latter is
not present in the config file

Signed-off-by: Shyamsundar Ranganathan <srangana@redhat.com>

### CODE REVIEW

**Review by raghavendra-talur — COMMENTED** [review]
The policy related changes look good to me. I am a bit worried about adding velero field to the secretsUtil as it doesn't really belong there.

Shouldn't we send the velero details from the ramenconfig in the methods which are implemented by secretUtil?

File: controllers/util/secrets_util.go [hunk 1]
```diff
@@ -230,7 +239,41 @@ func newS3ConfigurationSecret(s3SecretRef corev1.SecretReference, targetns strin
...
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "v" + s3SecretRef.Name,
+			Namespace: velero.Namespace,
+		},
+		Data: map[string]string{
+			velero.KeyName: "[default]\n" +
```
Comment (ShyamsundarR) [inline controllers/util/secrets_util.go:253]: This should change to:
```
'{{ (printf "default\n  aws_access_key_id = %s\n  aws_secret_access_key = %s"
  ((lookup "v1" "Secret" "ramen-system" "ramen-s3-secret-dr1").data.AWS_ACCESS_KEY_ID | base64dec)
  ((lookup "v1" "Secret" "ramen-system" "ramen-s3-secret-dr1").data.AWS_SECRET_ACCESS_KEY | base64dec)
) | base64enc }}'
```

The current method of processing the data does not work. The method above would copy a secret from the managed cluster, from one namespace to another, with the required format changes for velero. (So `ramen-system` as the namespace would actually be dr-cluster namespace on the managed cluster)

Comment (ShyamsundarR) [reply controllers/util/secrets_util.go]: Further to this, had to split it into 2 policies, as 2 secrets where one depends on the other being present, in the same policy does not work.

**Comment by ShyamsundarR** [comment]
> The policy related changes look good to me. I am a bit worried about adding velero field to the secretsUtil as it doesn't really belong there.
> 
> Shouldn't we send the velero details from the ramenconfig in the methods which are implemented by secretUtil?

Should be addressed in the latest set of changes. This part is good for a review.

**Comment by ShyamsundarR** [comment]
TODO:

- [ ] Testing with imperative apps in e2e, to ensure changes are in working order
- [ ] Any updates to default ramen config, to not configure veleroNamespaceSecretKeyRef
- [ ] Updates to drenv, to ignore creating velero secrets in the required format and namespace

File: controllers/util/secrets_util.go [hunk 2]
```diff
@@ -230,7 +281,28 @@ func newS3ConfigurationSecret(s3SecretRef corev1.SecretReference, targetns strin
 	}
 }
 
-func newConfigurationPolicy(name string, object runtime.RawExtension) *cpcv1.ConfigurationPolicy {
+func newVeleroSecret(s3SecretRef corev1.SecretReference, fromNS, veleroNS, keyName string) *localSecret {
+	return &localSecret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
```
Comment (nirs) [inline controllers/util/secrets_util.go:294]: Why not use `corev1.Secret{'? than we don't need to specify the TypeMeta.

I guess you keep the existing style to minimize the change?

File: controllers/util/secrets_util.go [hunk 3]
```diff
@@ -230,7 +281,28 @@ func newS3ConfigurationSecret(s3SecretRef corev1.SecretReference, targetns strin
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "v" + s3SecretRef.Name, // Disambiguate with a "v" in case fromNS and veleroNS are the same
```
Comment (nirs) [inline controllers/util/secrets_util.go]: This should use the constant, but inventing a new name we control will make this easier to work with.

File: controllers/util/secrets_util.go [hunk 4]
```diff
@@ -230,7 +281,28 @@ func newS3ConfigurationSecret(s3SecretRef corev1.SecretReference, targetns strin
...
+			Namespace: veleroNS,
+		},
+		Data: map[string]string{
+			keyName: "{{ (printf \"default\\n  aws_access_key_id = %s\\n  aws_secret_access_key = %s\" " +
+				"((lookup \"v1\" \"Secret\" \"" + fromNS +
+				"\" \"" + s3SecretRef.Name + "\").data.AWS_ACCESS_KEY_ID | base64dec) " +
+				"((lookup \"v1\" \"Secret\" \"" + fromNS +
+				"\" \"" + s3SecretRef.Name + "\").data.AWS_SECRET_ACCESS_KEY | base64dec)" +
+				") | base64enc }}",
```
Comment (nirs) [inline controllers/util/secrets_util.go:305]: This policy syntax is very hard to work with. Can we avoid it?

Or at lest prepare this as a multiline string to we can get everything lined up nicely and avoid the extra quoting?

File: controllers/util/secrets_util.go [hunk 5]
```diff
@@ -534,18 +640,26 @@ func (sutil *SecretsUtil) ensureS3SecretResources(secretName, namespace string)
...
-	if len(secretName)+len(namespace)+len(".") > policyNameLengthLimit {
+	if len(secretName)+len(namespace)+len(".")+formatPrefixLen > policyNameLengthLimit {
 		return fmt.Errorf("secret namespace.name (%s.%s) length exceeds maximum character limit (%d)",
 			secretName, namespace, policyNameLengthLimit)
 	}
 
-	secret, err := sutil.ensureS3SecretResources(secretName, namespace)
+	if format == SecretFormatVelero && veleroNS == "" {
+		return fmt.Errorf("requested format (%s) requires a target namespace", SecretFormatVelero)
+	}
```
Comment (nirs) [inline controllers/util/secrets_util.go:665]: Why we don't need this check for ramen secret?

File: controllers/util/secrets_util.go [hunk 6]
```diff
@@ -534,18 +640,26 @@ func (sutil *SecretsUtil) ensureS3SecretResources(secretName, namespace string)
...
-	return nil, sutil.deletePolicyResources(&secret, namespace)
+	return nil, sutil.deletePolicyResources(&secret, namespace, format)
 }
 
-func (sutil *SecretsUtil) AddSecretToCluster(secretName, clusterName, namespace, targetns string) error {
-	sutil.Log.Info("Add Secret", "cluster", clusterName, "secret", secretName)
+func (sutil *SecretsUtil) AddSecretToCluster(
+	secretName, clusterName, namespace, targetNS string,
+	format TargetSecretFormat,
+	veleroNS string,
```
Comment (nirs) [inline controllers/util/secrets_util.go:654]: accepting targetNS and velroNS is not consistent.

Is verlroNS another variant of targetNS?

File: controllers/util/secrets_util_test.go [hunk 7]
```diff
@@ -490,4 +495,83 @@ var _ = Describe("Secrets_Util", func() {
...
+					"")).To(Succeed())
+			})
+			It("Protects the secret with a finalizer", func() {
+				Expect(finalizerPresent(secretNames[0], util.SecretFormatRamen)).Should(BeTrue())
+			})
+			It("Creates an associated policy for the secret including the cluster", func() {
+				Expect(plRuleContains(plRuleName[0], tstNamespace, clusterNames[:1])).Should(BeTrue())
+				Expect(policyContains(policyName[0], tstNamespace, secrets[0])).Should(BeTrue())
+			})
+		})
```
Comment (nirs) [inline controllers/util/secrets_util_test.go:519]: The commit message promise to test that secret distribution works but only test that we create a policy for the secret. Maybe reword the commit message?

File: controllers/util/secrets_util_test.go [hunk 8]
```diff
@@ -490,4 +495,83 @@ var _ = Describe("Secrets_Util", func() {
...
+		When("Secret is created in the Velero namespace", func() {
```
Comment (nirs) [inline controllers/util/secrets_util_test.go:520]: We protect any secret in velero namespace?

File: controllers/util/secrets_util_test.go [hunk 9]
```diff
@@ -490,4 +495,83 @@ var _ = Describe("Secrets_Util", func() {
...
+					util.SecretFormatVelero,
+					veleroNS)).To(Succeed())
+			})
+			It("Protects the secret with an additional finalizer", func() {
+				Expect(finalizerPresent(secretNames[0], util.SecretFormatVelero)).Should(BeTrue())
+			})
+			It("Creates an associated policy for the secret including the cluster", func() {
+				Expect(plRuleContains(plRuleNameV[0], tstNamespace, clusterNames[:1])).Should(BeTrue())
+				Expect(policyContains(policyNameV[0], tstNamespace, secrets[0])).Should(BeTrue())
+			})
```
Comment (nirs) [inline controllers/util/secrets_util_test.go:536]: So we adopt any secret we find in velero namesapce and create a policy for the secret?

**Review by nirs — COMMENTED** [review]
The ramenctl changes looks good.

The first commit is huge and I don't fully understand it. Chang of this size is likely to contain bugs and regressions, I hope it is not targeted to 4.16.

Main issues:
- Why do we use the "v" prefix instead of creating a new clear ramen control name for the secret we generate?
- Why we need different finalizer for velero secerts? (not of code changes for this, not clear what is the benefit)
- I hope we don't adopt *any* secret from velero namespace. I think we should protect only secrets labeled with ramen specific label. This will help to avoid issue like #1434

File: controllers/util/secrets_util.go [hunk 10]
```diff
@@ -54,12 +72,45 @@ type SecretsUtil struct {
...
-		fmt.Sprintf(secretResourceNameFormat, secretPlRuleBaseName, secret),
-		fmt.Sprintf(secretResourceNameFormat, secretConfigPolicyBaseName, secret)
+	var pName string
+
+	switch format {
+	case SecretFormatRamen:
+		pName = ramenFormatPrefix + secret
+	case SecretFormatVelero:
+		pName = veleroFormatPrefix + secret
+	}
```
[the remark this answers is not in the document]
Comment (ShyamsundarR) [reply controllers/util/secrets_util.go:84]: The only reasonable thing to do is assert (panic), so that devs can catch a new format that is not handled in the future, or return and handle errors across all the functions using these internal routines.

Currently not adding a panic here for assertion as this is quite controlled in use.

**Comment by ShyamsundarR** [comment]
> The ramenctl changes looks good.
> 
> The first commit is huge and I don't fully understand it. Chang of this size is likely to contain bugs and regressions, I hope it is not targeted to 4.16.

It is for 4.16

> 
> Main issues:
> 
>     * Why do we use the "v" prefix instead of creating a new clear ramen control name for the secret we generate?

To control overall name length in these cases (including the name length of the OCM policy)

> 
>     * Why we need different finalizer for velero secerts? (not of code changes for this, not clear what is the benefit)

Answered inline

> 
>     * I hope we don't adopt _any_ secret from velero namespace. I think we should protect only secrets labeled with ramen specific label. This will help to avoid issue like [Ramen controllers are caching secret/configmaps from all namespaces #1434](https://github.com/RamenDR/ramen/issues/1434)

No we do not, we deliver a secret to the Velero namespace and this is not related to the posted issue.

File: controllers/util/secrets_util.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (BenamarMk) [reply controllers/util/secrets_util.go:84]: While the call is controlled in use, adding a `default` case would help prevent unintended changes to the function argument, especially since a `string` argument can be easily modified without notice. Additionally, `VSCode`'s autocomplete feature can occasionally introduce unexpected changes found only during testing.

**Review by BenamarMk — APPROVED** [review]
LGTM

**Review by raghavendra-talur — APPROVED** [review]

File: controllers/util/secrets_util.go (hunk 3 above)
Comment (ShyamsundarR) [reply controllers/util/secrets_util.go]: This is done

File: controllers/util/secrets_util.go (hunk 4 above)
Comment (ShyamsundarR) [reply controllers/util/secrets_util.go:314]: Added a comment with the desired output that it should generate
